JavaScript/chapter02/p01_remove_dups/avc278.js

```diff
@@ -20,14 +20,14 @@ const arrayToLinkedList = (arr) => {
 
 const compareLinkedLists = (A, B) => {
   if (!A && !B) return true;
-  if (!A || !B) return false;
 
   while (A !== null && B !== null) {
     if (A.val !== B.val) return false;
     A = A.next;
     B = B.next;
   }
-  return true;
+
+  return !((A && !B) || (!A && B));
```

Comment:
this could be simplified through [de Morgan's laws](https://en.wikipedia.org/wiki/De_Morgan%27s_laws):

```javascript
   !((A && !B) || (!A && B))
=> !(A && !B) && !(!A && B)
=> (!A || B) && (A || !B)
```

Apply a bit of [Distributivity](https://en.wikipedia.org/wiki/Distributive_property):

```javascript
   (!A || B) && (A || !B)
=> ((!A || B) && A) || ((!A || B) && !B)
=> (A && (!A || B)) || (!B && (!A || B))
=> (A && !A || A && B) || (!B && !A || !B && B)
=> (A && B) || (!B && !A)
```

This final one is another way of expressing the [Logical biconditional](https://en.wikipedia.org/wiki/Logical_biconditional):

```javascript
   (A && B) || (!B && !A)
=> (A !== null) === (B !== null)
```

based on the fact that the code is no longer in the loop in L24-L28, it means that it _has_ to be the case that `!(A !== null && B !== null) => A !== null || B !== null`. This means that the only way in which the previous expression can be true is

```javascript
return (A !== null) && (B !== null);
```

which is a more concise way of saying:

```javascript
if (!((A !== null) && (B !== null))) return false;
return true;
```

which in turn takes us back to:

```javascript
if (!A || !B) return false;
return true;
```

which is the reason why I mentioned that L23 was the solution! (although in prouction code `return (A !== null) && (B !== null);` may be preferred since it's simpler and the intention of "return true if BOTH linked lists have finished" is easier to grasp by humans).

(how did i get to all this result? i knew the answer upfront, so i cheated a bit since I was able to work my way backwards from the expected answer, BUT if i had not known the answer, one technique is to evaluate the expression with all possible boolean values and print out the logic table. look that table up in https://en.wikipedia.org/wiki/Truth_table#Binary_operations to see what's the immediate simplification. note that sometimes it's better to choose a suboptimal simplification if it allows the code flow to be easier to understand, especially if techniques like the [guard clause pattern](https://refactoring.com/catalog/replaceNestedConditionalWithGuardClauses.html) is used).

Reply:
Very interesting. Thank you for the in-depth explanation